File: OCP/MungedSolution/MarsRoverKata/Projectile.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace MarsRoverKata
{
    public class Projectile
    {
        public Projectile(Mars mars, bool isMortar)
        {
            Mars = mars;
            IsMortar = isMortar;
        }

        private Mars Mars { get; set; }
        private static readonly Dictionary<Direction, Point> PositionalAdjustments = new Dictionary<Direction, Point>() 
        { 
            { Direction.North, new Point(0, 1) },
            { Direction.South, new Point(0, -1) },
            { Direction.East, new Point(1, 0) },
            { Direction.West, new Point(-1, 0) }
        };

        public bool IsMortar { get; set; }
        private int MaxRange { get { return IsMortar ? 20 : 10; } }

        public void Launch(Direction facing, Point location)
        {
            bool collidedWithTarget = false;
            int moveIndex = 0;
            Point target = location;
            Point desiredPosition;
            while (!collidedWithTarget && moveIndex < MaxRange)
            {
                desiredPosition = CreateDesiredPosition(1, facing, target);
                target = CalculateProjectileFinalPosition(target, desiredPosition);
                collidedWithTarget = IsCollisionDetected(target);
                moveIndex++;
            }
            HitTarget(target);
        }

        private Point CreateDesiredPosition(int adjustmentFactor, Direction facing, Point location)
        {
            var adjustment = PositionalAdjustments[facing] * adjustmentFactor;
            return location + adjustment;
        }

        private Point CalculateProjectileFinalPosition(Point from, Point desired)
        {
            Point newDestination = desired;
            newDestination = CalculatePositionY(desired, newDestination);
            newDestination = CalculatePositionX(desired, newDestination);

            return newDestination;
        }

        private bool IsCollisionDetected(Point desired)
        {
            if (!IsMortar)
            {
                Point newDestination = desired;
                newDestination = CalculatePositionY(desired, newDestination);
                newDestination = CalculatePositionX(desired, newDestination);
                var obstacle = FindObstacle(newDestination);

                if (obstacle != null && obstacle.IsDestructable)
                {
                    return true;
                }

                return false;
            }
            return false;
        }

        private Point CalculatePositionX(Point desired, Point newDestination)
        {
            if (desired.X > Mars.Bounds.Width)
            {
                newDestination = new Point(0, desired.Y);
            }
            if (desired.X < 0)
            {
                newDestination = new Point(Mars.Bounds.Width, desired.Y);
            }
            return newDestination;
        }

        private Point CalculatePositionY(Point desired, Point newDestination)
        {
            if (desired.Y > Mars.Bounds.Height)
            {
                newDestination = new Point(desired.X, 0);
            }
            if (desired.Y < 0)
            {
                newDestination = new Point(desired.X, Mars.Bounds.Height);
            }
            return newDestination;
        }

        private Obstacle FindObstacle(Point point)
        {
            return Mars.Obstacles.SingleOrDefault(x => x.Location.Equals(point));
        }

        private void HitTarget(Point point)
        {
            var obstacle = FindObstacle(point);

            if (obstacle != null && obstacle.IsDestructable)
            {
                DestroyObstacle(obstacle);
            }
            else if (obstacle == null)
            {
                CreateObstacle(point);
            }
        }

        private void DestroyObstacle(Obstacle obstacle)
        {
            Mars.RemoveObstacle(obstacle);
        }

        private void CreateObstacle(Point point)
        {
            Obstacle obstacle = new Obstacle(point);
            Mars.AddObstacle(obstacle);
        }
    }
}

File: OCP/Java_Exercise/MarsRoverKata/src/main/java/MarsRoverKata/Projectile.java
package MarsRoverKata;

import com.google.common.base.Predicate;
import com.google.common.collect.Collections2;
import com.google.common.collect.Iterables;

import java.util.HashMap;
import java.util.Map;

public class Projectile {

    private static Map<Direction, Point> positionalAdjustments = new HashMap<Direction, Point>() {{
        put(Direction.North, new Point(0, 1));
        put(Direction.South, new Point(0, -1));
        put(Direction.East, new Point(1, 0));
        put(Direction.West, new Point(-1, 0));
    }};

    private Mars mars;
    private int maxRange;
    public boolean IsMortar;

    protected Projectile(Mars mars, boolean isMortar) {
        this.mars = mars;
        this.IsMortar = isMortar;
    }

    private int getMaxRange()
    {
        if(this.IsMortar)
            return 10;
        else
            return 5;
    }

    public void launch(Direction facing, Point location) {
        boolean collidedWithTarget = false;
        int moveIndex = 0;
        Point target = location;
        while (!collidedWithTarget && moveIndex < getMaxRange()) {
            Point desiredPosition = createDesiredPosition(1, facing, target);
            target = calculateProjectileFinalPosition(target, desiredPosition);
            collidedWithTarget = isCollisionDetected(target);
            moveIndex++;
        }
        hitTarget(target);
    }

    private static Point createDesiredPosition(int adjustmentFactor, Direction facing, Point location) {
        Point adjustment = Point.multiply(positionalAdjustments.get(facing), (adjustmentFactor));
        return Point.add(location, adjustment);
    }

    private Point calculateProjectileFinalPosition(Point from, Point desired) {
        Point newDestination = desired;
        newDestination = calculatePositionY(desired, newDestination);
        newDestination = calculatePositionX(desired, newDestination);

        return newDestination;
    }

    protected boolean isCollisionDetected(Point desired){
        if(!this.IsMortar) {
            Point newDestination = desired;
            newDestination = calculatePositionY(desired, newDestination);
            newDestination = calculatePositionX(desired, newDestination);
            Obstacle obstacle = findObstacle(newDestination);

            return obstacle != null && obstacle.isDestructable();
        }
        return false;
    }

    protected Point calculatePositionX(Point desired, Point newDestination) {
        if (desired.getX() > mars.getBounds().getWidth()) {
            newDestination = new Point(0, desired.getY());
        }
        if (desired.getX() < 0) {
            newDestination = new Point(mars.getBounds().getWidth(), desired.getY());
        }
        return newDestination;
    }

    protected Point calculatePositionY(Point desired, Point newDestination) {
        if (desired.getY() > mars.getBounds().getHeight()) {
            newDestination = new Point(desired.getX(), 0);
        }
        if (desired.getY() < 0) {
            newDestination = new Point(desired.getX(), mars.getBounds().getHeight());
        }
        return newDestination;
    }

    protected Obstacle findObstacle(final Point point) {
        return Iterables.getOnlyElement(Collections2.filter(mars.getObstacles(), new Predicate<Obstacle>() {
            @Override
            public boolean apply(Obstacle input) {
                return input.getLocation().equals(point);
            }
        }),null);
    }

    private void hitTarget(Point point) {
        Obstacle obstacle = findObstacle(point);

        if (obstacle != null && obstacle.isDestructable()) {
            destroyObstacle(obstacle);
        } else if (obstacle == null) {
            createObstacle(point);
        }
    }

    private void destroyObstacle(Obstacle obstacle) {
        mars.removeObstacle(obstacle);
    }

    private void createObstacle(Point point) {
        Obstacle obstacle = new Obstacle(point, false);
        mars.addObstacle(obstacle);
    }
}

